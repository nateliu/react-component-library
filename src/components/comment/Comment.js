import React from 'react';
import PropTypes from 'prop-types';

export default class Comment extends React.Component {
    constructor () {
        super();
        this.state = {timeString: ''};
    }

    componentWillMount() {
        this._updateTimeString()
        this._timer = setInterval(this._updateTimeString.bind(this),5000);
    }

    componentWillUnmount () {
        clearInterval(this._timer);
    }

    handleDeleteComment () {
        if (this.props.onDeleteComment) {
        this.props.onDeleteComment(this.props.index)
        }
    }

    render(){
        const { comment } = this.props
        return (
            <div className='comment'>
                <div className='comment-user'>
                    <span>{comment.userName} :&nbsp;</span>
                </div>
                <p dangerouslySetInnerHTML={{
                __html: this._getProcessedContent(comment.content)
                }} />
                <span className='comment-createdtime'>
                    {this.state.timeString}
                </span>
                <span className='comment-delete' onClick={this.handleDeleteComment.bind(this)}>Delete</span>
            </div>
        );
    }

    _updateTimeString () {
        const comment = this.props.comment
        const duration = (+Date.now() - comment.createdTime) / 1000
        this.setState({
        timeString: duration > 60
            ? `${Math.round(duration / 60)} Mins ago`
            : `${Math.round(Math.max(duration, 1))} Secs ago`
        })
    }

    _getProcessedContent (content) {
        return content
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;")
        .replace(/'/g, "&#039;")
        .replace(/`([\S\s]+?)`/g, '<code>$1</code>')
  }

}

Comment.propTypes = {
    comment: PropTypes.object.isRequired,
    onDeleteComment: PropTypes.func,
    index: PropTypes.number
}