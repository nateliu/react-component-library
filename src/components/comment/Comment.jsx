import React from 'react';
import PropTypes from 'prop-types';
import './Comment.css';

export default class Comment extends React.Component {
    constructor() {
        super();
        this.state = { timeString: '' };
    }

    componentWillMount() {
        this._updateTimeString()
        this._timer = setInterval(this._updateTimeString.bind(this), 5000);
    }

    componentWillUnmount() {
        clearInterval(this._timer);
    }

    handleDeleteComment() {
        if (this.props.onDeleteComment) {
            this.props.onDeleteComment(this.props.index)
        }
    }

    render() {
        const { comment } = this.props
        return (
            <div className='comment'>
                <div className='comment-user'>
                    <span>{comment.userName} :&nbsp;</span>
                </div>
                <p dangerouslySetInnerHTML={{
                    __html: this._getProcessedContent(comment.content)
                }} />
                <span className='comment-createdtime'>
                    {this.state.timeString}
                </span>
                <span className='comment-delete' onClick={this.handleDeleteComment.bind(this)}>Delete</span>
            </div>
        );
    }

    _updateTimeString() {
        const comment = this.props.comment
        const duration = (+Date.now() - comment.createdTime) / 1000
        this.setState({
            timeString: this._parseTimeString(duration)
        })
    }

    _parseTimeString(duration) {
        let result = '';
        const dayVariant = 60 * 60 * 24;
        const minVariant = 60;
        let singleNumber = 1;
        if (duration > dayVariant) {
            singleNumber = Math.round(duration / dayVariant);
            if (singleNumber > 1) {
                result = `${singleNumber} Days ago`
            } else {
                result = `${singleNumber} Day ago`
            }
        } else if (duration > minVariant) {
            singleNumber = Math.round(duration / minVariant);
            if (singleNumber > 1) {
                result = `${singleNumber} Mins ago`
            } else {
                result = `${singleNumber} Min ago`
            }
        } else {
            singleNumber = Math.round(Math.max(duration, 1));
            if (singleNumber > 1) {
                result = `${singleNumber} Secs ago`
            } else {
                result = `${singleNumber} Sec ago`
            }
        }
        return result;
    }

    _getProcessedContent(content) {
        return content
            .replace(/&/g, "&amp;")
            .replace(/</g, "&lt;")
            .replace(/>/g, "&gt;")
            .replace(/"/g, "&quot;")
            .replace(/'/g, "&#039;")
            .replace(/`([\S\s]+?)`/g, '<code>$1</code>')
    }

}

Comment.propTypes = {
    comment: PropTypes.object.isRequired,
    onDeleteComment: PropTypes.func,
    index: PropTypes.number
}