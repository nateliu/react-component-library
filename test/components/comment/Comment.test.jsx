import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Comment from '../../../src/components/comment/Comment';

function setup() {
    const props = {
        onDeleteComment: jest.fn(),
        comment: { "userName": "fakeName", "content": "comment in testing." },
        index: 1
    }

    const enzymeWrapper = shallow(<Comment {...props} />)

    return {
        props,
        enzymeWrapper
    }
}

describe('comment/Comment', () => {
    it('renders correctly with props', () => {
        const { enzymeWrapper } = setup();
        expect(shallowToJson(enzymeWrapper)).toMatchSnapshot();
    })
})
