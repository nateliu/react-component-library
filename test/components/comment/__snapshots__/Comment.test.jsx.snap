// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment/Comment renders correctly with props 1`] = `
<div
  className="comment"
>
  <div
    className="comment-user"
  >
    <span>
      fakeName
       :Â 
    </span>
  </div>
  <p
    dangerouslySetInnerHTML={
      Object {
        "__html": "comment in testing.",
      }
    }
  />
  <span
    className="comment-createdtime"
  >
    NaN Sec ago
  </span>
  <span
    className="comment-delete"
    onClick={[Function]}
  >
    Delete
  </span>
</div>
`;
